WIP Performance Work

jimi-c opened the pull request:
This is an experimental branch for testing performance related improvements.

Depends-On: https://github.com/ansible/ansible-zuul-jobs/pull/245

jimi-c committed:
Rework how the core engine works (nbd...)

jimi-c committed:
Adding in sivel's debug cprofile decorator

jimi-c committed:
Adding in process affinity code from mitogen

jimi-c committed:
Make host serialization smarter and some other tweaks

jimi-c committed:
More tweaks to use ParseResultDottedDict

jimi-c committed:
Revert worker side and plugins to use dotted notation

Also fixes copy() in dotted dict class we're using, because before it
was reverting to a normal dict() when it was copied.

jimi-c committed:
Actually pass task_vars via file (no hostvars yet though)

Also be sure to squash objects before serializing

jimi-c committed:
Use msgpack for (de)serializing task_vars and add hostvars back

jimi-c committed:
Additional tweaks

* Improving ssh selector and finite state machine performance
* Removing use of results thread and processing results directly
  from the final_q

jimi-c committed:
Tweak to callbacks to allow for a dotted dict to work as a play

jimi-c committed:
Adding a new plugin loader and additional tweaks

jimi-c committed:
Fixing clean_vars to not use connection_loader.all

* Uses new PluginFinder code and removes a lot of inefficient code
  like the re.compile to speed things up
* Fixes a bug with the logic

jimi-c committed:
Pre-calculate the repr of the valid attrs for base objects

jimi-c committed:
Optimize iterator use for strategies

Because we have already done the work via get_next_task_for_host using
peek=True, we don't need to redo the exact same work. Instead we just
save the calculated host state back to the iterator to make the state
official.

jimi-c committed:
Ensure the basedir in the dataloader is always an absolute path

jimi-c committed:
Improvements to combine_vars

* Added a flag to allow the validation step to be skipped, which
  can be useful when you know for sure you're combining mutable mappings
  and don't have to validate the objects.
* Added a flag to allow the merge to occur in place without first making
  a copy of the destination mapping.
* Changed combine_vars use in get_vars() to use both flags for speed.

jimi-c committed:
Tweaking vars plugins use of inventory source directories

jimi-c committed:
Don't compile regex inside clean_vars (speed improvement)

jimi-c committed:
Add option to cprofile decorator to not clean up temp files

Useful for saving the output pstat files for use with graphing tools.

jimi-c committed:
Tweaks to playbook objects

* Add an option to init to not generate an id. This is useful for
  copy actions where we don't need a new id.
* Cache dumped attributes. This saves a lot of hasattr calls once
  the attrs have been dumped. Doing a set on any attribute invalidates
  the cached attrs and they'll be re-dumped next pass.

jimi-c committed:
Use new PluginFinder in TaskExecutor for connections

jimi-c committed:
Tweaks to vars manager

* Create one templar and re-use it as necessary.
* Cache host/group vars per host to avoid re-reading them.

jimi-c committed:
Fix bug introduced in base var plugin class

jimi-c committed:
Fixing closing order of multiprocessing stuff

jimi-c committed:
Tweak how dumped attributes are cached/invalidated

jimi-c committed:
Folding test architecture into perf_idea

jimi-c committed:
Fixing problem after rebase from devel

jimi-c committed:
Testing with no affinity

ci_complete

jimi-c committed:
Cleaning up some field attribute stuff

jimi-c committed:
Cleaning up tests

jimi-c committed:
Fixing handler notifications

jimi-c committed:
Disabling two relative path tests for DataLoader

One of the performance improvements from perf_idea was to make the DataLoader
class init with the basedir unfracked. This prevents a LOT of os.path calls
later, but it does mean the DataLoader will never deal with relative paths in
methods like the dwim searches. Overall this should be better as it removes a
potential bug with relative pathing, but it does have implications for testing.

These tests were disabled because they're probably not required anymore. They
could be shifted to test against the $CWD of the tests but we may not care.

dw commented:
Rather than send task_vars down to worker (massive single-threaded O(config) generation+serialization bottleneck, and in this case filesystem IO -- good luck with NFS), send template-expanded values up from worker to parent (fully parallelized generation and basically free, aside from fiddly code). Take care not to block task scheduling loop while waiting for processed values

jimi-c commented:
@dw the problem is we need the task_vars in the main proc for a few templated fields (the task name and now the throttle value), so we have to generate them there. I did test just throwing that away and having the worker regenerate them, however that was actually quite a bit slower. Actually just typing that I realize now that since I've moved callbacks to the result side we don't need to template the task name but the throttle field stands :/

dw commented:
Given throttle is inherently 1 for each host-task combination (Ansible runs no more than one task per host simultaneously), does it make sense for its template to have host-task binding? Even if host binding is desired, IIRC there is precedent elsewhere to template using only the first matched host, although this might be something from an ancient version. For 1000 hosts, eliminating 999 get_vars() per task would be a big deal, especially with a slow vars configuration.

The binding of throttle isn't documented, only that it can be set on a per-task basis, so making it play+task binding and calculating it once might be something that could live in the changelog, if mentioned at all. There is also the option (as for run_once) of verifying throttle is a template before calling get_vars().

For callbacks I use a buffer where events are reordered to ensure on_task_start for the first host is always reported before subsequent callbacks, by delaying them until the worker has reported the template-expanded name. That way plugins always see the 'correct' event ordering

Regardless of strategy the important thing is that dreaded get_vars() call, there is no escaping its blast radius

jimi-c commented:
@dw I will give that a try, though it's a little more complicated for non-linear strategies (though we can cache the task_vars and look them up per uuid or something). I had originally wanted to move the get_vars call to the worker side, but earlier in my testing it was slower than sending them over via file. I'll try it again though as only generating vars once for the (play, task) context instead of per-host may help.

jimi-c commented:
Also, with static worker sets, we'd have to ship facts/other vars includes to each worker, which may eat up some of the gain. As is, keeping the call to get_vars in the main proc simplifies things.

dw commented:
I have a replacement vars manager that solves your facts problem and quite a bit more.. it's something that should be upstream. My experiences with fielding proposals and PRs has been abysmal to this point, primarily because most of your communication does not happen in open spaces. Is there some other way I could approach contributing this to you?

jimi-c committed:
Cleaning up more unit tests

jimi-c committed:
Cleaning up sanity tests and fixing a found bug in handlers

jimi-c committed:
Fixing more problems in tests

* Temporarily disabled direct use of send_callback as there are serialization
  issues with several of the objects being sent across the Queues

jimi-c committed:
Fixing more problems

* args was not set to always post validate?

pabelanger commented:
recheck

jimi-c committed:
Fully removing the old plugin loader

jimi-c committed:
More cleanup

* Fixing botmeta since we replaced the old plugin loader
* Fixing role duplication detection and handling
* Cleaning up pep and other sanity issues

jimi-c committed:
Fixing plugin namespace creating for test/filter plugins

jimi-c committed:
Re-enabling set_plugin_options for shell in TaskExecutor

jimi-c committed:
Make plugin scanning smarter about plugin types/extensions

This also cleans up some other code, which was used to set the preferred
extension type based on the connection plugin.

jimi-c committed:
Cleaning up action plugins

jimi-c committed:
Add more role data into serialized task

* Modifies include_vars to use new role data fields

jimi-c committed:
More cleanup

* Fixing templating of loop control fields
* Feeding include_vars info back to workers via updates

jimi-c committed:
Fixing problem with any_errors_fatal

jimi-c committed:
Work to clean up tests
